src/engine/client/cl_cgame.cpp

```diff
@@ -1474,74 +1447,65 @@ void CGameVM::QVMSyscall(int index, Util::Reader& reader, IPC::Channel& channel)
 		// All keys
 
 		case CG_KEY_GETCATCHER:
-			IPC::HandleMsg<Key::GetCatcherMsg>(channel, std::move(reader), [this] (int& catcher) {
+			IPC::HandleMsg<Keyboard::GetCatcherMsg>(channel, std::move(reader), [this] (int& catcher) {
 				catcher = Key_GetCatcher();
 			});
 			break;
 
 		case CG_KEY_SETCATCHER:
-			IPC::HandleMsg<Key::SetCatcherMsg>(channel, std::move(reader), [this] (int catcher) {
+			IPC::HandleMsg<Keyboard::SetCatcherMsg>(channel, std::move(reader), [this] (int catcher) {
 				Key_SetCatcher(catcher);
 			});
 			break;
 
-		case CG_KEY_GETKEYNUMFORBINDS:
-			IPC::HandleMsg<Key::GetKeynumForBindsMsg>(channel, std::move(reader), [this] (int team, const std::vector<std::string>& binds, std::vector<std::vector<int>>& result) {
+		case CG_KEY_GETKEYSFORBINDS:
+			IPC::HandleMsg<Keyboard::GetKeysForBindsMsg>(channel, std::move(reader), [this] (int team, const std::vector<std::string>& binds, std::vector<std::vector<Keyboard::Key>>& result) {
 				for (const auto& bind : binds) {
-					result.push_back({});
-					for (int i = 1; i < Util::ordinal(keyNum_t::MAX_KEYS); i++) {
-						if (i >= 'A' && i <= 'Z') {
-							continue; // Skip since these are equivalent to lowercase letters.
-						}
-						char buffer[MAX_STRING_CHARS];
-
-						Key_GetBindingBuf(i, team, buffer, MAX_STRING_CHARS);
-						if (bind == buffer) {
-							result.back().push_back(i);
-							continue;
-						}
-					}
+					result.push_back(Keyboard::GetKeysBoundTo(team, bind));
 				}
 			});
 			break;
 
-		case CG_KEY_KEYNUMTOSTRINGBUF:
-			IPC::HandleMsg<Key::KeyNumToStringMsg>(channel, std::move(reader), [this] (int keynum, std::string& result) {
-				result = Keyboard::KeyToString(Keyboard::Key::FromLegacyInt(keynum));
+		case CG_KEY_GETCHARFORSCANCODE:
+			IPC::HandleMsg<Keyboard::GetCharForScancodeMsg>(channel, std::move(reader), [this] (int scancode, int& result) {
+				result = Keyboard::GetCharForScancode(scancode);
+				if (!result) {
+					// Not sure if this fallback is ever useful. Usually SDL falls back on QWERTY itself.
+					result = Keyboard::ScancodeToAscii(scancode);
+				}
 			});
 			break;
 
 		case CG_KEY_SETBINDING:
-			IPC::HandleMsg<Key::SetBindingMsg>(channel, std::move(reader), [this] (int keyNum, int team, std::string cmd) {
-				Keyboard::SetBinding(Keyboard::Key::FromLegacyInt(keyNum), team, std::move(cmd));
+			IPC::HandleMsg<Keyboard::SetBindingMsg>(channel, std::move(reader), [this] (Keyboard::Key key, int team, std::string cmd) {
+				Keyboard::SetBinding(key, team, std::move(cmd));
```

Comment:
Is any validation necessary for key?

Reply:
Good idea, I added some validation here as well as in the deserialization.